src/operator/contrib/sync_batch_norm-inl.h

```diff
@@ -0,0 +1,562 @@
+enum BatchNormOpOutputs {kOut, kMean, kVar};
+enum BatchNormOpAuxiliary {kMovingMean, kMovingVar};
+enum BatchNormBackResource {kTempSpace};
+}  // namespace syncbatchnorm
+
+struct SyncBatchNormParam : public dmlc::Parameter<SyncBatchNormParam> {
+  float eps;
+  float momentum;
+  bool fix_gamma;
+  bool use_global_stats;
+  bool output_mean_var;
+  int ndev;
+  std::string key;
+  DMLC_DECLARE_PARAMETER(SyncBatchNormParam) {
+    DMLC_DECLARE_FIELD(eps).set_default(1e-3f)
+    .describe("Epsilon to prevent div 0");
+    DMLC_DECLARE_FIELD(momentum).set_default(0.9f)
+    .describe("Momentum for moving average");
+    DMLC_DECLARE_FIELD(fix_gamma).set_default(true)
+    .describe("Fix gamma while training");
+    DMLC_DECLARE_FIELD(use_global_stats).set_default(false)
+    .describe("Whether use global moving statistics instead of local batch-norm. "
+              "This will force change batch-norm into a scale shift operator.");
+    DMLC_DECLARE_FIELD(output_mean_var).set_default(false)
+    .describe("Output All,normal mean and var");
+    DMLC_DECLARE_FIELD(ndev).set_default(1)
+      .describe("The count of GPU devices");
+    DMLC_DECLARE_FIELD(key)
+      .set_default("")
+      .describe("Hash key for synchronization");
+  }
+};
+
+// Modified from https://github.com/brucechin/SharedTensor
+template<class T>
+class SharedND {
+ private:
+  int nDev;
+  T mean;
+  T *data;
+  bool *flag;
+  bool meanReady = false;
+  bool meanInited = false;
+  std::mutex mutex_;
+
+ public:
+  explicit SharedND(int ndev) :nDev(ndev) {
+      flag = new bool[ndev];
+      data = new T[ndev];
+      memset(flag, false, ndev * sizeof(bool));
+  }
+
+  ~SharedND() {
+    delete [] flag;
+    delete [] data;
+  }
+
+  bool Push(T input, int index) {
+    if (flag[index] == false) {
+      data[index] = input;
+      flag[index] = true;
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  T Pop(int index) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    while (!MeanReady()) {}
+    flag[index] = false;
+    T tmp = mean;
+    ResetMean();
+    return tmp;
+  }
+
+  bool MeanReady() {
+    if (meanReady) {
+      return true;
+    }
+    for (int i = 0; i < nDev; i++) {
+      if (!flag[i]) {
+        return false;
+      }
+    }
+    for (int i = 1; i < nDev; i++) {
+      data[0] += data[i];
+    }
+    if (!meanInited) {
+      mean = mshadow::NewTensor<cpu, real_t>(data[0].shape_, 0.0f);
+      meanInited = true;
+    }
+    mean = data[0] * 1.0f /  nDev;
+    meanReady = true;
+    return true;
+  }
+
+  void ResetMean() {
+    for (int i = 0; i < nDev; i++) {
+      if (flag[i]) return;
+    }
+    meanReady = false;
+  }
+};
+
+template<class T>
+class GlobalShared {
+ public:
+  T* Register(const std::string &key, int ndev) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    auto it = registry_.find(key);
+    if (it != registry_.end()) return it->second;
+    T *newT = new T(ndev);
```

Comment:
memory is not released pointed by these raw pointers

Reply:
addressed in the class deconstruction function :)